cpp/src_prims/cache/cache_util.h

```diff
@@ -0,0 +1,342 @@
+                                 int *cache_idx, int n_cache_sets,
+                                 int *cache_time, int time, int *out_idx) {
+  int block_offset = blockIdx.x * associativity;
+
+  const int items_per_thread = ceildiv(associativity, nthreads);
+
+  // the size of rank limits how large associativity can be used in practice
+  __shared__ int rank[items_per_thread * nthreads];
+  rank_set_entries<nthreads, associativity>(cache_time, n_cache_sets, rank);
+
+  // Each thread will fill items_per_thread items in the cache.
+  // It uses a place, only if it was not udated at the current time step
+  // (cache_time != time).
+  // We rank the places acconding to the time stamp, least recently used
+  // elements come to the front.
+  // We fill the least recently used elements with the working set.
+  // there might be elements which cannot be assigned to cache loc.
+  // these elements are assigned -1.
+
+  for (int j = 0; j < items_per_thread; j++) {
+    int i = threadIdx.x + j * nthreads;
+    int t_idx = block_offset + i;
+    bool mask = (i < associativity);
+    // whether this slot is available for writing
+    mask = mask && (cache_time[t_idx] != time);
+
+    // rank[i] tells which element to store by this thread
+    // we look up where is the idx stored in the input array
+    if (mask) {
+      int k = find_nth_occurrence(cache_set, n, blockIdx.x, rank[i]);
+      mask = mask && k > -1;
+      if (k > -1) {
+        int idx_val = in_idx[k];
+        cache_idx[t_idx] = idx_val;
+        out_idx[k] = t_idx;
+        cache_time[t_idx] = time;
+      }
+    }
+  }
+}
+
+/**
+ * @brief Get the cache indices for vectors stored in the cache.
+ *
+ * For every index in idx, we look up the corresponding cache position.
+ * If idx[k] is stored in the cache, then is_cached[k] is set to true, and
+ * out_cache_idx[k] stores the corresponding cache idx.
+ *
+ * If idx[k] is not stored in the cache, then we assign a cache set to it.
+ * This idx is returned in out_cache_idx[k], and is_cached[k] is set to false.
+ * In this case AssignCacheIdx should be called, to get an assigned position
+ * within the cache set.
+ *
+ * Cache_time is assigned to the time input argument for all elements in idx.
+ *
+ * @param [in] idx array of vector indices that we want to look up in the cache, size [n]
+ * @param [n] number of indices to look up
+ * @param [inout] vec_idx indices stored in the cache, size [n_cache_sets x associativity]
+ * @param [in] n_cache_sets number of cache sets
+ * @param [in] associativity number of vectors in cache set
+ * @param [inout] cache_time time stamp when the indices were cached, size [n_cache_sets * associativity]
+ * @param [out] out_cache_idx cache indices of the working set elements size [n]
+ * @param [out] is_cached  whether the element is cached size[n]
+ * @param [in] time iteration counter (used for time stamping)
+ */
+inline __global__ void get_cache_idx(int *idx, int n, int *vec_idx,
```

Comment:
A minor issue: the `inline` specifier is used here to indicate that there could be multiple definitions of this non-template function, when the header is included in multiple translation units (otherwise we would get multiple definition error during linking). AFAIK this is a valid use case for the `inline` specifier.

The compiler emits the following `warning: inline qualifier ignored for "__global__" function`. This is understood and can be ignored, since we do not intend the global function to be inlined. 

If the warning message is disturbing, then we have the following options: 
 - add the following compiler option `-Xcudafe="--diag_suppress=2943" `. (In general one can display the error/warning message numbers with the `-Xcudafe="--display_error_number"` flag, and use the diag_suppress option to disable it)
 - use `static` keyword instead of `inline`
 - put the function in an unnamed namespace

Let me know which option would you prefer.

Reply:
I'd prefer unnamed namespace.